pire/extra/count.h

```diff
@@ -297,9 +302,205 @@ class AdvancedCountingScanner : public BaseCountingScanner<AdvancedCountingScann
 
 	friend class Impl::ScannerGlueCommon<AdvancedCountingScanner>;
 	friend class Impl::CountingScannerGlueTask<AdvancedCountingScanner>;
+	friend AdvancedCountingScanner Impl::MakeAdvancedCountingScanner<AdvancedCountingScanner>(const Fsm&, const Fsm&, bool*);
 };
 
-}
+class NoGlueLimitCountingScanner : public BaseCountingScanner<NoGlueLimitCountingScanner> {
+public:
+	using ActionIndex = ui32;
+	using TActionsBuffer = std::unique_ptr<ActionIndex[]>;
+
+	class State {
+	public:
+		size_t Result(int i) const { return ymax(m_current[i], m_total[i]); }
+
+	private:
+		InternalState m_state;
+		TVector<ui32> m_current;
+		TVector<ui32> m_total;
+
+		friend class NoGlueLimitCountingScanner;
+
+#ifdef PIRE_DEBUG
+		yostream& operator << (yostream& s, const State& state)
+		{
+			s << state.m_state << " ( ";
+			for (size_t i = 0; i < state.m_current.size(); ++i)
+				s << state.m_current[i] << '/' << state.m_total[i] << ' ';
+			return s << ')';
+		}
+#endif
+	};
+
+private:
+	TActionsBuffer ActionsBuffer;
+	const ActionIndex* Actions = nullptr;
+
+public:
+	NoGlueLimitCountingScanner() = default;
+	NoGlueLimitCountingScanner(const Fsm& re, const Fsm& sep, bool* simple = nullptr);
+	NoGlueLimitCountingScanner(const NoGlueLimitCountingScanner& rhs) : BaseCountingScanner(rhs) {
+		if (rhs.ActionsBuffer) {
+			Y_ASSERT(rhs.Actions);
+			ActionsBuffer = TActionsBuffer(new ActionIndex [*rhs.Actions]);
+			memcpy(ActionsBuffer.get(), rhs.ActionsBuffer.get(), *rhs.Actions * sizeof(ActionIndex));
+			Actions = ActionsBuffer.get();
+		} else {
+			Actions = rhs.Actions;
+		}
+	}
+
+	NoGlueLimitCountingScanner(NoGlueLimitCountingScanner&& other) : BaseCountingScanner() {
+		Swap(other);
+	}
+
+	NoGlueLimitCountingScanner& operator=(NoGlueLimitCountingScanner rhs) {
+		Swap(rhs);
+		return *this;
+	}
+
+	void Swap(NoGlueLimitCountingScanner& s) {
+		LoadedScanner::Swap(s);
+		DoSwap(ActionsBuffer, s.ActionsBuffer);
+		DoSwap(Actions, s.Actions);
+	}
+
+	void Initialize(State& state) const
+	{
+		state.m_state = m.initial;
+		state.m_current.assign(RegexpsCount(), 0);
+		state.m_total.assign(RegexpsCount(), 0);
+	}
+
+	////////////////////////////////////////////////////////////
+	// Copy-past from BaseCountingScanner because of changed State type
+	// Alternative is to make State template parameter in all these functions in BaseCountingScanner
```

Comment:
Perhaps, it's possible to parametrize `BaseCountingScanner` by State?

Reply:
To do this, we need to move State definitions from inner scopes of corresponding Scanners to outer scope. Is it ok?